packages/vm/src/runTx.ts

```diff
@@ -33,6 +33,21 @@ export async function runTx(this: VM, opts: RunTxOpts): Promise<RunTxResult> {
   // create a reasonable default if no block is given
   opts.block = opts.block ?? Block.fromBlockData({}, { common: opts.tx.common })
 
+  if (opts.skipHardForkValidation !== true) {
+    if (opts.tx.common.hardfork() !== this._common.hardfork()) {
+      // If hardforks aren't same then we can posibily try upgrading tx hardfork but it may
+      // be fraught with challenges. Better to just reject the tx and the tx sender can
+      // update the tx as per new hardfork
+      const msg = _errorMsg('tx has a different hardfork than the vm', this, opts.block, opts.tx)
+      throw new Error(msg)
+    }
+    if (opts.block._common.hardfork() !== this._common.hardfork()) {
+      // Block and VM's hardfork should match as well
+      const msg = _errorMsg('block has a different hardfork than the vm', this, opts.block, opts.tx)
+      throw new Error(msg)
+    }
+  }
```

Comment:
I was first irritated by this new `skipHardForkValidation` option but now that I am getting is and see how this is applied I find this a super-great solution to guard the current behavior under this flag and at the same time to allow to go narrower and deactivate, eventually one by one respectively over time! 🤩

Really nice, since we have talked about this a lot already but didn't dare yet to just plainly introduce these checks, this is a super nice and gentle way to finally do! 🙏 👍

Reply:
yea! i was also apprehensive to add this but then realized it was a more progressive way to go for our tests! glad you feel the same :slightly_smiling_face: